pkg/dump/dump.go

```diff
@@ -45,175 +46,303 @@ const (
-	case Full:
-		dumpFull("request", kind, method, request, 0)
-	}
+// run runs the dumping goroutine of the dumper.
+func (d *dumper) run() {
+	go func() {
+		for req := range d.q {
+			if req.dir != nop {
+				method := methodName(req.method)
+				d.RLock()
+				detail, ok := d.details[method]
+				if !ok {
+					detail = d.rules.detailOf(method)
+				}
+				d.RUnlock()
+				switch detail {
+				case Name:
+					d.name(req.dir, req.kind, method, req.msg, req.latency)
+				case Full:
+					d.full(req.dir, req.kind, method, req.msg, req.latency)
+				}
+			}
+			if req.sync != nil {
+				close(req.sync)
+			}
+		}
+	}()
 }
 
-// ReplyMessage dumps a CRI reply.
-func ReplyMessage(kind, name string, reply interface{}, latency time.Duration) {
-	method := name[strings.LastIndex(name, "/")+1:]
-	opt.Lock()
-	defer opt.Unlock()
-	switch opt.verbosityOf(method) {
-	case NameOnly:
-		dumpName("reply", kind, method, reply, latency)
-	case Full:
-		dumpFull("reply", kind, method, reply, latency)
-	}
+// sync waits until all the persent messages in the queue are dumped.
+func (d *dumper) sync() {
+	ch := make(chan struct{}, 1)
+	dump.q <- &dumpreq{dir: nop, sync: ch}
+	_ = <-ch
 }
 
-func dumpName(dir, kind, method string, msg interface{}, latency time.Duration) {
-	go func() {
-		switch dir {
-		case "request":
+// configure (re)configures dumper
+func (d *dumper) configure(o *options) {
+	d.Lock()
+	defer d.Unlock()
+
+	d.debug = o.Debug
+	d.rules = o.rules.duplicate()
+
+	if d.path != o.File || d.disabled != o.Disabled {
+		if d.file != nil {
+			log.Info("closing old message dump file %q...", d.path)
+			d.file.Close()
+			d.file = nil
+		}
+		d.disabled = o.Disabled
+
+		if d.disabled {
 			return
-		case "reply":
-			if _, ok := msg.(error); ok {
-				dumpWarn(dir, latency, "(%s) FAILED %s: %vs", kind, method, msg.(error))
-			} else {
-				dumpLine(dir, latency, "(%s) REQUEST %s", kind, method)
+		}
+
+		d.path = o.File
+		if d.path != "" {
+			var err error
+			log.Info("opening new message dump file %q...", d.path)
+			d.file, err = os.Create(d.path)
+			if err != nil {
+				log.Error("failed to open file %q: %v", d.path, err)
 			}
 		}
-	}()
+	}
+
+	d.train(nil)
 }
 
-func dumpFull(dir, kind, method string, msg interface{}, latency time.Duration) {
-	go func() {
-		switch dir {
-		case "request":
+// train trains the dumper with the given set of messages.
+func (d *dumper) train(names []string) {
+	if names != nil {
+		d.methods = make([]string, len(names), len(names))
+	} else {
+		names = d.methods
+	}
+	d.details = make(map[string]level)
+	for idx, name := range names {
+		method := methodName(name)
+		detail := d.rules.detailOf(method)
+		log.Info("%s: %v", method, detail)
+		d.methods[idx] = method
+		d.details[method] = detail
+	}
+}
+
+// name does a name-only dump of the given message.
+func (d *dumper) name(dir direction, kind, method string, msg interface{}, latency time.Duration) {
+	switch dir {
+	case request:
+		return
```

Comment:
We don't dump requests?

Reply:
With name-only dumping we dump request/reply-pairs as a single line, and only once the reply has come in, so we know whether the request succeeded or failed.